Issue #7064: Added influenceFormat property in SuppressWithPlainTextCommentFilter

AbhinavRajputEXE opened the pull request:
Issue #7064

nmancus1 commented:
@AbhinavRajputEXE please share errors and what you have tried so far to fix them.

AbhinavRajputEXE commented:
@nmancus1 
> @AbhinavRajputEXE please share errors

this is the error: [comment](https://github.com/checkstyle/checkstyle/issues/7064#issuecomment-1019561761)

>and what you have tried so far to fix them.

For some time I didn't know what this error means but after some digging I found out that I need to make some changes in the test file of SuppressWithPlainTextCommentFilter too for the new property InfluenceFormat and to understand it, I was studying old related PRs and trying to understand the code so that I can understand what changes I need to make.
If you can help me a little here to understand what changes I need to make, that would be really awesome. I've worked on the test files before and I know that the test file contains the line and column number of violations in Input files but apart from that, I don't know much about them.

nmancus1 commented:
From https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=7465&view=logs&j=c902ebb4-c9f8-5f09-4e17-ff78fbbc842e&t=9ca98c81-ff64-58f0-9d03-a23ac1c4a111&l=812:

```
[ERROR] testAllCheckSectionJavaDocs  Time elapsed: 2.254 s  <<< ERROR!
java.lang.Error: Error was thrown while processing src/main/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithPlainTextCommentFilter.java
...
Caused by: com.google.common.truth.ComparisonFailureWithFacts: 
SuppressWithPlainTextCommentFilter's class-level JavaDoc
diff (-expected +actual):
    @@ -109,12 +109,6 @@
     <pre>
     &lt;module name="Checker"&gt; &lt;property name="fileExtensions" value="sql"/&gt; &lt;module name="SuppressWithPlainTextCommentFilter"&gt; &lt;property name="offCommentFormat" value="@cs-\: ([\w\|]+)"/&gt; &lt;property name="checkFormat" value="$1"/&gt; &lt;/module&gt; &lt;/module&gt;
     </pre>
    -<p>
    -Example of how to configure the check to suppress more than one checks. The influence format format is specified in the second regexp group.
    -</p>
     <pre>
    -&lt;module name="SuppressWithPlainTextCommentFilter"&gt; &lt;property name="commentFormat" value="@cs-\: ([\w\|]+) influence (\d+)"/&gt; &lt;property name="checkFormat" value="$1"/&gt; &lt;property name="influenceFormat" value="$2"/&gt; &lt;/module&gt;
    -</pre>
    -<pre>
     -- @cs-: RegexpSinglelineCheck -- @cs-: FileTabCharacterCheck CREATE TABLE STATION ( ID INTEGER PRIMARY KEY, CITY CHAR(20), STATE CHAR(2), LAT_N REAL, LONG_W REAL);
     </pre>


```

```
...
[ERROR] testAllXmlExamples  Time elapsed: 1.09 s  <<< ERROR!
com.puppycrawl.tools.checkstyle.api.CheckstyleException: 
config_filters.xml has invalid Checkstyle xml (cannot initialize module SuppressWithPlainTextCommentFilter - Property 'commentFormat' does not exist, please check the documentation): <module name="SuppressWithPlainTextCommentFilter">
    <property name="commentFormat" value="@cs-\: ([\w\|]+) influence (\d+)"/>
    <property name="checkFormat" value="$1"/>
    <property name="influenceFormat" value="$2"/>
...

```
```
...
[ERROR] Failures: 
[ERROR]   SuppressWithPlainTextCommentFilterTest.testEqualsAndHashCodeOfSuppressionClass:376 Error: EqualsVerifier found a problem in class com.puppycrawl.tools.checkstyle.filters.SuppressWithPlainTextCommentFilter$Suppression.
-> Significant fields: equals does not use firstLine, or it is stateless.

For more information, go to: https://www.jqno.nl/equalsverifier/errormessages
value of: isSuccessful()
expected to be true
...

```

There are three failures/ errors.

1. There is a difference between the Javadoc in the class you have modified, and it's corresponding documentation in `config_filters.xml` file. They must be the same.
2. `Property 'commentFormat' does not exist, please check the documentation` - this property does not exist.
3. Link is provided in error message, please read about failure there.

AbhinavRajputEXE commented:
@nmancus1 
I accidentally added that property before but now I've Removed it as SupressWithPlainWhiteTextCommentFilter doesn't support that property.

AbhinavRajputEXE commented:
Hey @nmancus1 
is there any documentation related to test files, after reading them and running debug on a few tests I kinda understood it but I'm not completely sure what it is doing exactly except that it is chacking the input files and the violations in those files?

can you please tell me briefly what exactly is happening there and is there any common format that these test files follow?

Like in the test below I understood that it is checking violation messages in the `InputSuppressWithPlainTextCommentFilterWithDefaultCfg.java` file and has the line and col. numbers of violations in that file that are related to tabs are there in this test.
```
@Test
    public void testFilterWithDefaultConfig() throws Exception {
        final DefaultConfiguration filterCfg =
            createModuleConfig(SuppressWithPlainTextCommentFilter.class);

        final DefaultConfiguration checkCfg = createModuleConfig(FileTabCharacterCheck.class);
        checkCfg.addProperty("eachLine", "true");

        final String[] suppressed = {
            "5:7: " + getCheckMessage(FileTabCharacterCheck.class, MSG_FILE_CONTAINS_TAB),
            "10:1: " + getCheckMessage(FileTabCharacterCheck.class, MSG_CONTAINS_TAB),
        };

        final String[] violationMessages = {
            "5:7: " + getCheckMessage(FileTabCharacterCheck.class, MSG_FILE_CONTAINS_TAB),
            "8:7: " + getCheckMessage(FileTabCharacterCheck.class, MSG_CONTAINS_TAB),
            "10:1: " + getCheckMessage(FileTabCharacterCheck.class, MSG_CONTAINS_TAB),
        };

        verifySuppressed(
            "InputSuppressWithPlainTextCommentFilterWithDefaultCfg.java",
            removeSuppressed(violationMessages, suppressed),
            filterCfg, checkCfg
        );
    }
```

AbhinavRajputEXE commented:
@nmancus1 
Now I'm getting this error on `mvn clean verify`
What could be causing this error😅
and sorry if I'm asking a bit too many doubts😅😅

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.452 s <<< FAILURE! - in com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest
[ERROR] testAllCheckSectionJavaDocs  Time elapsed: 1.451 s  <<< ERROR!
java.lang.Error: Error was thrown while processing src\main\java\com\puppycrawl\tools\checkstyle\filters\SuppressWithPlainTextCommentFilter.java
        at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.examineCheckSection(XdocsJavaDocsTest.java:192)
        at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.testAllCheckSectionJavaDocs(XdocsJavaDocsTest.java:165)
Caused by: com.google.common.truth.ComparisonFailureWithFacts:
SuppressWithPlainTextCommentFilter's class-level JavaDoc
diff (-expected +actual):
    @@ -109,12 +109,6 @@
     <pre>
     &lt;module name="Checker"&gt; &lt;property name="fileExtensions" value="sql"/&gt; &lt;module name="SuppressWithPlainTextCommentFilter"&gt; &lt;property name="
offCommentFormat" value="@cs-\: ([\w\|]+)"/&gt; &lt;property name="checkFormat" value="$1"/&gt; &lt;/module&gt; &lt;/module&gt;
     </pre>
    -<p>
    -Example of how to configure the check to suppress more than one checks. The influence format format is specified in the second regexp group.
    -</p>
     <pre>
    -&lt;module name="SuppressWithPlainTextCommentFilter"&gt; &lt;property name="checkFormat" value="$1"/&gt; &lt;property name="influenceFormat" value="$2"/&gt; &
lt;/module&gt;
    -</pre>
    -<pre>
     -- @cs-: RegexpSinglelineCheck -- @cs-: FileTabCharacterCheck CREATE TABLE STATION ( ID INTEGER PRIMARY KEY, CITY CHAR(20), STATE CHAR(2), LAT_N REAL, LONG_W 
REAL);
     </pre>
        at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.examineCheckSection(XdocsJavaDocsTest.java:192)
        at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.testAllCheckSectionJavaDocs(XdocsJavaDocsTest.java:165)

```

AbhinavRajputEXE commented:
@nmancus1 

can you help me with this error?😅
what does it mean by  `Significant fields: equals does not use firstLine, or it is stateless.`

```
[ERROR] Tests run: 22, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.261 s <<< FAILURE! - in com.puppycrawl.tools.checkstyle.filters.SuppressWithPlainTextCommentFilterTest
[ERROR] testEqualsAndHashCodeOfSuppressionClass  Time elapsed: 0.06 s  <<< FAILURE!
com.google.common.truth.AssertionErrorWithFacts: 
Error: EqualsVerifier found a problem in class com.puppycrawl.tools.checkstyle.filters.SuppressWithPlainTextCommentFilter$Suppression.
-> Significant fields: equals does not use firstLine, or it is stateless.

For more information, go to: https://www.jqno.nl/equalsverifier/errormessages
value of: isSuccessful()
expected to be true
	at com.puppycrawl.tools.checkstyle.filters.SuppressWithPlainTextCommentFilterTest.testEqualsAndHashCodeOfSuppressionClass(SuppressWithPlainTextCommentFilterTest.java:376)

```

AbhinavRajputEXE commented:
Hey, @nmancus1 @romani 
can you review the changes please and help me with [this(comment)](https://github.com/checkstyle/checkstyle/pull/11248#issuecomment-1026213681).

AbhinavRajputEXE commented:
@nmancus1 
I introduced 2 variables `lastLine` and `firstLine` for the `influenceFormat` property and when I add these in the equal function it gives me an error.

```
[ERROR] [checkstyle] [ERROR] C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\ci\src\main\java\com\puppycrawl\tools\checkstyle\filters\SuppressWithPlainTextCommentFilter.java:666:9: Cyclomatic Complexity is 11 (max allowed is 10). [CyclomaticComplexity]
```

can I increase the cyclomatic complexity limit for this file or is there any other way to fix it?

romani commented:
You have 8 commit, please squash all in one.

AbhinavRajputEXE commented:
@romani 

> Split method is two methods

I need to split `equals` and `hashCode` into two?..right?😅

AbhinavRajputEXE commented:
> Split method is two methods

@romani 

I split the equals method into two because this was causing the high cyclomatic complexity error.
But now I'm getting these errors.
```
[ERROR] [checkstyle] [ERROR] C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\ci\src\main\java\com\puppycrawl\tools\checkstyle\filters\SuppressWithPlainTextCommentFilter.java:682:9: Missing a Javadoc comment. [MissingJavadocMethod]
[ERROR] [checkstyle] [ERROR] C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\ci\src\main\java\com\puppycrawl\tools\checkstyle\filters\SuppressWithPlainTextCommentFilter.java:682:49: '{' is not preceded with whitespace. [WhitespaceAround]
[ERROR] [checkstyle] [ERROR] C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\ci\src\main\java\com\puppycrawl\tools\checkstyle\filters\SuppressWithPlainTextCommentFilter.java:684:13: Boolean expression complexity is 8 (max allowed is 7). [BooleanExpressionComplexity]
```

I fixed the middle one but
it's still saying high complexity for the method I created in the last error, how should I change it so that it does not give me this error.
And what comments should I add for the first error?

romani commented:
I still see 2 commit

AbhinavRajputEXE commented:
> I still see 2 commit

squashed.

romani commented on src/xdocs/config_filters.xml:
```diff
@@ -2296,6 +2296,14 @@ public class UserService {
+                  lines preceding/at/following the suppression comment.</td>
+              <td><a href="property_types.html#String">String</a></td>
+              <td><code>"0"</code></td>
+              <td>5.0</td>
```
Put here 10.0 for now it is our next version.
This version define in what version this property appears.
So users that use old version should clearly see it

romani reviewed (changes requested):
I do not see update for tests. Please extra tests with coverage of new feature.
Please review pitest failure it will guide you how to improve testing or code even further. Whole CI should be green. Please revit all failures and fix them

AbhinavRajputEXE commented:
@romani 
> I do not see update for tests. Please extra tests with coverage of new feature.

Yes, I actually kept it for the end.. and,
as I asked [earlier(comment)](https://github.com/checkstyle/checkstyle/pull/11248#issuecomment-1026213681) can you please tell me a little bit more about test files or if is there any documentation.
Here is what I understood about them till now.

_In a test, we create a module configuration and we add properties to that config.
and then we create two strings supressedViolationMessages and expectedViolationMessages
and then there are the line and col. number of the violations of the Input files in those string_

Is this correct? and I did not understand the last part of the test where we call `verifyFilterWithInlineConfigParser` method
which contains the file name of the Input file. what this method is doing exactly, what is the Input file in this.

AbhinavRajputEXE commented:
> Please review pitest failure it will guide you how to improve testing or code even further. Whole CI should be green. Please revit all failures and fix them

`Error:  Failed to execute goal org.pitest:pitest-maven:1.7.4:mutationCoverage (default-cli) on project checkstyle: Mutation score of 99 is below threshold of 100 -> [Help 1]`

'It says mutation score is below the threshold
Is it because I have not created tests for influenceFormat?

romani commented:
Yes, let's start testing

romani commented:
>Is this correct?

Yes 
Test method are creating check and filter from config that is placed at top of input file. It execute config without filter and with filter and make sure that filter remove violation.

AbhinavRajputEXE commented:
@romani 

I've written this one test, I took a little bit of help from `SuppressWithNearbyCommentFilter.java` file.
Altho it's not complete yet because I have got a few doubts.

1. In try block verifySuppressed take _"Input file name with extention"_ as its first argument. So how do I know which InputFile name I have to add here for this test? and
2. For now I've added `CommonUtil.EMPTY_STRING_ARRAY;` as an example in suppressed and violationMessages string array. But How do I know what line and column number I have to add from which Input file for this test.

Do I need to select an Input file manually that fits this test and have to add line and col. number manually for `suppressed` and `violationMessage` string array? or is there is any other way?

```
    @Test
    public void testInvalidInfluenceFormat() throws Exception {
        final DefaultConfiguration filterCfg =
            createModuleConfig(SuppressWithPlainTextCommentFilter.class);
        filterCfg.addProperty("influenceFormat", "a");

        final DefaultConfiguration checkCfg = createModuleConfig(FileTabCharacterCheck.class);
        checkCfg.addProperty("eachLine", "true");
        
        final String[] suppressed = CommonUtil.EMPTY_STRING_ARRAY;
        
        final String[] violationMessages = CommonUtil.EMPTY_STRING_ARRAY;
        
        try {
            verifySuppressed(" ", removeSuppressed(violationMessages, suppressed), filterCfg, checkCfg);
            assertWithMessage("CheckstyleException is expected").fail();
        }
        catch (CheckstyleException ex) {
            assertWithMessage("Invalid exception message")
                .that(ex)
                .hasCauseThat().hasMessageThat()
                .isEqualTo("unable to parse influence"
                        + " from 'SUPPRESS CHECKSTYLE MemberNameCheck' using a");
        }
    }
```

AbhinavRajputEXE commented:
@romani 
I added a test in the test file altho it's not complete.
Can you please help me understand [this(comment above)](https://github.com/checkstyle/checkstyle/pull/11248#issuecomment-1033045661)?

I'm getting this error BTW.

```
[ERROR] Failures: 
[ERROR]   SuppressWithPlainTextCommentFilterTest.testInvalidInfluenceFormat:292->verifySuppressed:703->AbstractModuleTestSupport.verify:287->AbstractModuleTestSupp
ort.verify:325->AbstractModuleTestSupport.verify:347->AbstractModuleTestSupport.verify:375 Files with expected violations and actual violations differ.
value of: keySet()
expected to be empty
but was : [C:\Users\abhin\Desktop\Checkstyle_Folder\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\filters\suppresswithplaintextcommentfilter\InputS
uppressWithPlainTextCommentFilterWithCustomOnAndOffComments.java]


```
I'll squash the commits after this.

AbhinavRajputEXE commented:
@romani 
I just need to know a little bit more about, how to choose Input files for our tests which we put in `verifySuppressed` method call argument, and how those lines and col. number works, which are in that string arrays like `suppressed`  and `violationMessages` string array in the test file,
like: `"6: " + getCheckMessage(RegexpSinglelineCheck.class, MSG_REGEXP_EXCEEDED,".*[a-zA-Z][0-9].*"` (like here 6 as line number)
Please tell me about this😅

nmancus1 commented:
@AbhinavRajputEXE it is best to approach PR like this in a way similar to how we create bug reports and show how documentation updates work:

1. Create new test input file
2. Create test method using desired config, test should be failing
3. Make changes in related class to satisfy test and reflect expected behavior.
4. Update documentation.

You can take a look at https://github.com/checkstyle/checkstyle/pull/11270/files as a good example of adding new test inputs and test methods with documentation update.

AbhinavRajputEXE commented:
> 1. Create new test input file
> 2. Create test method using desired config, test should be failing
> 3. Make changes in related class to satisfy test and reflect expected behavior.
> 4. Update documentation.

Ohk, so please correct me if I'm wrong.

1. I've to create a test method in the respective test class and configure that method by adding the properties to it.
2. And after this, I've to create an Input file for this method and make intentional violations for that test to fail, and put a violation comment in that violation line.
3. After creating the method and making an Input file for that method I've to run `mvn clean verify` and It'll give me errors that in this line and column number we have errors in the Input file.
4. And to fix those errors I've to give the line and column number in that test method's suppressed and violationMessages string array accordingly.

Am I right?😅
@nmancus1

nmancus1 commented:
@AbhinavRajputEXE this sounds about right, also see tests at `SuppressWithNearbyCommentFilterTest.java` for examples.

AbhinavRajputEXE commented:
@nmancus1 
I made some changes and it says `Failures: 0, Errors: 0, Skipped: 0` and gives me 2 warnings, and after the `BUILD FAILURE` message It gives me an error.

What is the meaning of these two warnings and that one error?
Please help.

```
[INFO] Tests run: 451, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] 
[INFO] --- modernizer-maven-plugin:2.3.0:modernizer (modernizer) @ checkstyle ---
[INFO] 
[INFO] --- maven-checkstyle-plugin:3.1.2:check (sevntu-checkstyle-check) @ checkstyle ---
[INFO] You have 0 Checkstyle violations.
[INFO]
[INFO] --- jacoco-maven-plugin:0.8.7:check (default-check) @ checkstyle ---
[INFO] Loading execution data file C:\Users\abhin\Desktop\Checkstyle_Folder\checkstyle\target\jacoco.exec
[INFO] Analyzed bundle 'checkstyle' with 799 classes
[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.filters.SuppressWithPlainTextCommentFilter.Suppression: lines covered ratio is 0.94, but expected
 minimum is 1.00
[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle.filters.SuppressWithPlainTextCommentFilter.Suppression: branches covered ratio is 0.97, but expec
ted minimum is 1.00
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  20:19 min
[INFO] Finished at: 2022-02-24T20:45:33+05:30
[INFO] ------------------------------------------------------------------------
```
Error after BUILD FAILURE message
```

[ERROR] Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.7:check (default-check) on project checkstyle: Coverage checks have not been met. See log for det
ails. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.7:check (default-check) on project checkstyle: Co
verage checks have not been met. See log for details.


```

romani commented:
https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations#how-to-generate-ut-coverage-report

please generate/review report of test coverage. Google about "code test coverage", educate yourself on what it means.
This failure means that you do not have enough test coverage.

romani commented:
Yes, all coverage we archive by Input files.
To make sure we do not do tests for situation that never happens in real life and we do know why each block of our code required and how to trigger it's execution.

AbhinavRajputEXE commented:
@romani 
I made a new test and an Input file for the InfluenceFormat property and I was expecting an error message regarding violation messages and their line and col. number after running `mvn clean verify -e`
But I got this message, what does this mean, It says about the path and name of my newly created Input file, But I don't exactly understand.

```
[ERROR] Failures: 
[ERROR]   SuppressWithPlainTextCommentFilterTest.testInfluenceFormat:299->verifySuppressed:703->AbstractModuleTestSupport.verify:287->AbstractModuleTestSupport.ver
ify:325->AbstractModuleTestSupport.verify:347->AbstractModuleTestSupport.verify:375 Files with expected violations and actual violations differ.
value of   : keySet()
missing (1): C:\Users\abhin\Desktop\Checkstyle_Folder\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\filters\suppresswithplaintextcommentfilter\Inpu
tSuppressWithPlainTextCommentFilterInfluenceFormat.java
---
expected   : [C:\Users\abhin\Desktop\Checkstyle_Folder\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\filters\suppresswithplaintextcommentfilter\Inp
utSuppressWithPlainTextCommentFilterInfluenceFormat.java]
but was    : []
[INFO]
[ERROR] Tests run: 3720, Failures: 1, Errors: 0, Skipped: 3
```

romani commented:
Run test in debug mode in IDE and see what is going on.

nick-mancuso commented:
@AbhinavRajputEXE ping, do you plan to continue this PR?

AbhinavRajputEXE commented:
@nick-mancuso I was focusing on understanding a gsoc project and drafting a proposal for it, i don't have any problem if anyone can do it.

romani commented:
@AbhinavRajputEXE, please make sure you have in your PRs fixes for first/second/third gsoc issues
